trie-db/src/trie_codec.rs

```diff
@@ -0,0 +1,635 @@
+///
+/// This function makes the assumption that all child references in an inline trie node are inline
+/// references.
+pub fn encode_compact<L>(db: &TrieDB<L>) -> Result<Vec<Vec<u8>>, TrieHash<L>, CError<L>>
+	where
+		L: TrieLayout
+{
+	// The stack of nodes through a path in the trie. Each entry is a child node of the preceding
+	// entry.
+	let mut stack: Vec<EncoderStackEntry<L::Codec>> = Vec::new();
+
+	// TrieDBNodeIterator guarantees that:
+	// - It yields at least one node.
+	// - The first node yielded is the root node with an empty prefix and is not inline.
+	// - The prefixes yielded are in strictly increasing lexographic order.
+	let iter = TrieDBNodeIterator::new(db)?;
+
+	// Following from the guarantees about TrieDBNodeIterator, we guarantee that after the first
+	// iteration of the loop below, the stack always has at least one entry and the bottom (front)
+	// of the stack is the root node, which is not inline. Furthermore, the iterator is not empty,
+	// so at least one iteration always occurs.
+	for item in iter {
+		match item {
+			Ok((prefix, node_hash, node)) => {
+				// Skip inline nodes, as they cannot contain hash references to other nodes by
+				// assumption.
+				if node_hash.is_none() {
+					continue;
+				}
+
+				// Unwind the stack until the new entry is a child of the last entry on the stack.
+				// If the stack entry prefix is a prefix of the new entry prefix, then it must be a
+				// direct parent as the nodes are yielded from the iterator in pre-order traversal
+				// order.
+				while let Some(mut last_entry) = stack.pop() {
+					if prefix.starts_with(&last_entry.prefix) {
+						// advance_child_index preconditions are satisfied because of iterator
+						// correctness.
+						last_entry.advance_child_index(&prefix)
+							.expect(
+								"all errors from advance_child_index indicate bugs with \
+								TrieDBNodeIterator or this function"
+							);
+						stack.push(last_entry);
+						break;
+					} else {
+						stack.last_mut()
+							.expect(
+								"the first entry on the stack is the root node from the iterator; \
+								the prefix of root node is an empty nibble vec; \
+								an empty nibble vec is a prefix of all other nibble slices; \
+								thus the root node entry must still be on the stack; \
+								qed"
+							)
+							.omit_child(last_entry)?;
+					}
+				}
+
+				let children_len = match node.node_plan() {
+					NodePlan::Empty | NodePlan::Leaf { .. } => 0,
+					NodePlan::Extension { .. } => 1,
+					NodePlan::Branch { .. } | NodePlan::NibbledBranch { .. } => NIBBLE_LENGTH,
+				};
+				stack.push(EncoderStackEntry {
+					prefix,
+					node,
+					child_index: 0,
+					omit_children: vec![false; children_len],
+					encoded_nodes: Vec::new(),
+					_marker: PhantomData::default(),
+				});
+			}
+			Err(err) => match *err {
+				// If we hit an IncompleteDatabaseError, just ignore it and continue encoding the
+				// incomplete trie.
```

Comment:
A quick additional word on the fact that we use it to process proof, and that for this use case the trie is inherently incomplete (to get the rational behind this skip).